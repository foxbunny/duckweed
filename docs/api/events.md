# `duckweed.events`

The duckweed events object exposes a few function that help us with handling and
transforming event objects.

## `duckweed.events.from()`

```javascript
(processor, eventHandler) => EventHandler
```

This is a helper function that composes the event processor and an event handler
(such as the one returned from the [`act()`](../act.md) function). Processors
will transform the event objects in various ways and the result will be added to
the argument list of the event handler.

The processor is a simple function that takes an event object and returns an
array of arguments. For example:

```javascript
const mouseEvent = (event) => [event.clientX, event.clentY];
```

The above example will transform the event object into two arguments which
represent the x and y coordinates of the mouse event. When used with an event
handler, it may look like this:

```javascript
const E = duckweed.events;

html("div", {on: {
  click: E.from(mouseEvent, (x, y) => alert(`You clicked at ${x}, ${y}`))
}});
```

By convention, we name the processor after the names of the events they handle.

## `duckweed.events.inputEvent()`

This processor will convert the event object into an input value.

## `duckweed.events.checkboxEvent()`

This processor will convert a change event on a checkbox into two arguments,
value of the `checked` property, and the input value.

## `duckweed.events.auto()`

This is an autopilot processor that includes all of the above and determines
which processor to run depending on the event and element type.

[Documentation index](../main.md) | [API reference index](./main.md)
