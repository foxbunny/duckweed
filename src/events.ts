/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import * as is from './is';

type ArgsProcessor = (...args) => any[] | undefined;
type EventHandler = (...args) => void;


/**
 * Decorates an event handler with a processor
 */
const from = (processor: ArgsProcessor, handler: EventHandler): EventHandler => (...eventArgs) =>
  (args => args && handler(...args)
  )(processor(...eventArgs));


/**
 * Processor that handles input DOM events
 */
const inputEvent: ArgsProcessor = (event: Event): [string] =>
  [(event.target as HTMLInputElement).value];


/**
 * Processor that handles change DOM events on checkboxes
 */
const checkboxEvent: ArgsProcessor = (event: Event): [boolean, string] =>
  [(event.target as HTMLInputElement).checked, (event.target as HTMLInputElement).value];


/**
 * Automatic processor that handles various events and hooks
 */
const auto = (...eventCallbackArgs) => {
  const first = eventCallbackArgs[0];
  if (is.vnode(first)) {
    // This is mostly for hooks. We add the vnode objects to args.
    return eventCallbackArgs;
  } else if (is.changeEvent(first) && is.checkbox(first.target)) {
    return checkboxEvent(first);
  } else if (is.inputEvent(first) && is.input(first.target)) {
    first.preventDefault();
    return inputEvent(first);
  } else if (is.pathData(first)) {
    return [first];
  }
  return [];
};


export {
  ArgsProcessor,
  EventHandler,
  from,
  inputEvent,
  checkboxEvent,
  auto,
};
