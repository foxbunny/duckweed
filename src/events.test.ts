/**
 * (c) 2017 Hajime Yamasaki Vukelic
 * All rights reserved.
 */

import * as snabbdom from 'snabbdom';
import h from 'snabbdom/h';
import eventlisteners from 'snabbdom/modules/eventlisteners';
import props from 'snabbdom/modules/props';

import * as e from './events';


const patch = snabbdom.init([eventlisteners, props]);


const scaffold = (elm: string, cfg: {[key: string]: any}) =>
  (vnode =>
    ((vnode.elm as any).vnode = vnode) && vnode.elm
  )(patch(document.createElement('div'), h(elm, cfg)));


describe('events', () => {
  it('Should handle input events', () => {
    const handler = jest.fn();
    const r = scaffold('input', {
      on: {input: e.from(e.inputEvent, handler)},
      props: {value: 'my input'},
    });
    r.dispatchEvent(new Event('input'));
    expect(handler).toHaveBeenCalledWith('my input');
  });

  it('Should handle checkbox events', () => {
    const handler = jest.fn();
    const r = scaffold('input', {
      on: {change: e.from(e.checkboxEvent, handler)},
      props: {value: 'my checkbox', checked: false, type: 'checkbox'},
    });
    r.dispatchEvent(new Event('change'));
    expect(handler).toHaveBeenCalledWith(false, 'my checkbox');
  });

  it('Should handle many events automatically', () => {
    const handler = jest.fn();
    const r1 = scaffold('input', {
      on: {
        input: e.from(e.auto, handler),
      },
      props: {value: 'my value'},
    });
    const r2 = scaffold('input', {
      on: {
        change: e.from(e.auto, handler),
      },
      props: {type: 'checkbox', value: 'my value', checked: true},
    });
    r1.dispatchEvent(new Event('input'));
    r2.dispatchEvent(new Event('change'));
    expect(handler.mock.calls).toMatchSnapshot();
  });

  it('Should return undefined if event processor returns undefined', () => {
    const handler = jest.fn();
    const processor = () => undefined;
    const f = e.from(processor, handler);
    expect(f()).toBeUndefined();
    expect(handler).not.toHaveBeenCalled();
  });
});
