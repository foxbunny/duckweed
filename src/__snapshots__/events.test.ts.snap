// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events Should handle many events automatically 1`] = `
Array [
  Array [
    "my value",
  ],
  Array [
    true,
    "my value",
  ],
]
`;
